src/OrchardCore.Modules/OrchardCore.Settings/SiteSettings.cs

```diff
@@ -1,26 +1,88 @@
+using System.Collections.Generic;
 using Microsoft.AspNetCore.Routing;
+using Newtonsoft.Json.Linq;
 using OrchardCore.Documents;
+using OrchardCore.Entities;
 
 namespace OrchardCore.Settings
 {
     // When updating class also update SiteSettingsDeploymentSource and SettingsStep.
     public class SiteSettings : DocumentEntity, ISite
     {
+        private JObject _properties = new();
+
+        public new JObject Properties
+        {
+            get => _properties;
+            set
+            {
+                _properties = value ?? new JObject();
```

Comment:
You could just not mutate `_properties` if the value is null in place of creating a new `JObject`, and then maybe not clear the cache.

I didn't have the time to try it on my side but are the `new Properties` well used in both way, for example populated from an existing `SiteSettings` in the database?

---

Yes, I never took the time to manage an `IsReadOnly` flag for a Document, but just for info, when we load a Document to be updated we call somewhere `.GetOrCreateMutableAsync()`, but when we get a cached Document (for sharing) we call somewhere `.GetOrCreateImmutableAsync()` which returns a document that normally should not be mutated.

Reply:
Is there a reason why to not set it to a new object? Yes there are many way to a a accomplish the same result. The base class sets it to a new Object and usually we use Put and As without having to worry if Properties is null. It is best to always initialize it.

Reply:
No problem to init it to a new `JObject`, here I meant something like this

        private JObject _properties = new();

        public new JObject Properties
        {
            get => _properties;
            set
            {
                if (value is not null)
                {
                    _properties = value;
                }

                _cache.Clear();
            }
        }


But here I think we could just use this as done in `Entity`.

    public new JObject Properties { get; set; } = new();

Not even sure that here we need a setter, so maybe

    public new JObject Properties { get; } = new();

Reply:
I think we have to use a setting since we are overriding a property that define a setter.

If we ignore `Properties = null;` the user will not be able to clear Properties. This will change the behavior. I think currently it's correct. Setting it to null will just initialize it to a new object that is ready to put/get settings.

Reply:
Okay for the setter.

In my example we still clear the cache, Hmm, if the value is null you could do `_properties.RemoveAll()` instead.

I would need to focus on it a little more but,

If we `Get` an immutable Document for sharing it should not be mutated, and if we `Load` a mutable Document for updating normally it should follow the current pattern to only update their related part in `Properties`.

So not sure we need to be able to clear this cache and then to override `Properties`.

Reply:
So, also not sure that we need a `Put` at this level, hmm unless if some updaters need to read a property while others update it, the entries would need to be cleared.

I talk about making a Document `IsReadOnly` or not, not so hard to do but at least we could do it for `SiteSettings`, meaning that in `GetSiteSettingsAsync()` the Document would be marked as `IsReadOnly` (internal and not serializable bool), here the cache would operate, and in `LoadSiteSettingsAsync()` we would set `IsReadOnly` to false, here the cache would not operate.

Reply:
I may introduce the support of this flag in the DocumentManager itself when I will have time.

Reply:
@jtkech do we need to wait until adding support for the DocumentManager? I don't think we should block this PR for this. I think the PR as is provides a valid value. Maybe we take this one and if you add `IsReadOnly` support we can modify what is needed then.